assets/search.js

```diff
@@ -1,20 +1,107 @@
-document.addEventListener("DOMContentLoaded", function () {
-  const searchElement = document.getElementById("search");
-  if (searchElement) {
-    searchElement.addEventListener("blur", function () {
-      searchElement.value = "";
-      searchElement.setAttribute("placeholder", "⌘K");
-    });
+import {
+  createChannel,
+  each,
+  main,
+  on,
+  resource,
+  sleep,
+  spawn,
+  suspend,
+} from "https://esm.sh/effection@4.0.0-alpha.5";
+
+await main(function* () {
+  const input = document.getElementById("search");
+  if (!input) {
+    console.log(`Search could not be setup because input was not found.`);
+    return;
+  }
+
+  const label = input.closest("label");
+  if (!label) {
+    console.log(
+      `Search could not be setup because label element was not found.`,
+    );
+    return;
+  }
+
+  const button = input.nextElementSibling;
+  if (!button) {
+    console.log(
+      `Search could not be setup because button element was not found.`,
+    );
+    return;
+  }
 
-    document.addEventListener("keydown", function (event) {
-      if (event.metaKey && event.key === "k") {
-        event.preventDefault();
-        searchElement.focus();
-        searchElement.removeAttribute("placeholder");
+  let lastBlur;
+  yield* forEach(
+    yield* join([
+      on(input, "focus"),
+      on(button, "focus"),
+      on(input, "blur"),
+      on(button, "blur"),
+    ]),
+    function* (event) {
+      if (event.type === "blur") {
+        lastBlur = yield* spawn(function* () {
+          yield* sleep(15);
+          input.value = "";
+          input.setAttribute("placeholder", "⌘K");
+          input.classList.remove("focused");
+        });
+      } else {
+        if (lastBlur) {
+          yield* lastBlur.halt();
+        }
+        input.removeAttribute("placeholder");
+        input.classList.add("focused");
       }
-      if (event.key === "Escape") {
-        searchElement.blur();
+    },
+  );
+
+  yield* forEach(on(document, "keydown"), function* (event) {
+    if (event.metaKey && event.key === "k") {
+      event.preventDefault();
+      input.focus();
+    }
+    if (event.key === "Escape") {
+      input.blur();
+    }
+  });
+
+  yield* suspend();
+});
+
+/**
+ * @param {Stream<Event, void>} stream
+ * @param {(event: Event) => Operation<void>} op
+ * @returns
+ */
+function forEach(stream, op) {
+  return spawn(function* () {
+    for (const event of yield* each(stream)) {
+      yield* op(event);
+      yield* each.next();
+    }
+    yield* suspend();
```

Comment:
I don't think you want to suspend here, because at this point the loops is done, and so the only thing it will do is hold onto the task.

Reply:
I'll try that.

Reply:
But I don't want it to end. It should stay connected to that node until teardown.

Reply:
If your task ever reaches line 85, then there is, by definition, no more events left in the stream. Is it that you want any subtasks and resources to stay alive until the `spawn()` is shut down? In that case, you might want to consider a `resource()` instead of a `spawn()`

Reply:
> if your task ever reaches line 85, then there is, by definition, no more events left in the stream.

That wasn't my understanding based on past experience. I thought that the for/each closed when it depleted events, not when the event stream source was closed. This resulted in the loop completing while there were still events possibly coming down from the source. I might be conflating scenarios, but that was my understanding of how it worked.

Reply:
A for-yield-each will close proceed when the subscription is fully consumed. I.e. the `{ done: true }` happens. For that reason, the for loop is actually a good primitive for synchronization, because it consumes the stream fully.

In fact, because the type of `on(evenTarget, eventName)` is `Stream<Event, never>` it will never actually reach the end because there is no end to the possibility of a user clicking, or typing, or whatever.

Reply:
@cowboyd Ok, I eliminated unnecessary suspend and dropped forEach. I'm using `all` in `join` and `for/each` everywhere. It's even more 😍 than it was before.